### PR TITLE
[MXNET-375] Lp Pooling and Global Lp Pooling

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
We follow the same implementation as PyTorch's implementation:
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L445-L465

### Essentials ###
- [x] The PR title starts with [MXNET-375]
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support Lp Pooling
- [x] Unit tests for Lp Pooling

## Comments ##
- For bridging the gap between MXNet and ONNX operators
Comparison of performance of Avgpool:
(before &nbsp;&nbsp;&nbsp;&nbsp; after)
1D Pooling on CPU:
0.039776878356933595&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.049250895977020266
2D Pooling on CPU:
0.018903429508209228&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.01837245225906372
3D Pooling on CPU:
0.9366579866409301&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.915347421169281
1D Pooling on GPU:
0.002552490234375&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.0024517107009887694
2D Pooling on GPU:
0.0012897467613220215&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.001228764057159424
3D Pooling on GPU:
0.048471152782440186&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0.048865692615509035
Script used:
```python
import mxnet as mx
from mxnet.test_utils import check_speed, rand_ndarray, rand_shape_nd

def get_params(dim):
    if dim == 1:
        return (50, 3, 20000), (16, ), (0, ), (2, )
    if dim == 2:
        return (50, 3, 45, 45), (16, 16), (0, 0), (2, 2)
    if dim == 3:
        return (50, 3, 32, 32, 32), (16, 16, 16), (0, 0, 0), (2, 2, 2)

def main():
    for n_dim in range(3, 6):
        test_shape, kernel, pad, stride = get_params(n_dim - 2)
        test_input = rand_ndarray(test_shape, 'default')
        data = mx.sym.Variable('data')
        sym = mx.sym.Pooling(data=data, kernel=kernel, pad=pad, stride=stride, pool_type='avg', cudnn_off=True)
        for context in [mx.cpu(), mx.gpu()]:
            print(check_speed(sym, location={'data':test_input}, ctx=context, N=100))


if __name__ == '__main__':
    main()
```

### CODE REVIEW

**Comment by haojin2** [comment]
@reminisce @rahul003 @anirudh2290 @eric-haibin-lin Please give a review if you have time, thanks!

**Comment by piiswrong** [comment]
Where does LP pool come from? we need to add documentations for it. See `describe` when registering Pooling.

**Comment by haojin2** [comment]
@piiswrong The first appearance of LpPool seems to be in this paper: Convolutional Neural Networks Applied to
House Numbers Digit Classification (https://arxiv.org/pdf/1204.3968.pdf). I'll update the doc with the link to the paper.

File: src/operator/nn/pool.cuh [hunk 1]
```diff
@@ -212,23 +213,22 @@ template <typename DType>
 __global__ void pool_sum_1d_gpu_kernel(const int nthreads, const DType* in_data, const int channels,
                                        const int width, const int pooled_width, const int kernel_w,
                                        const int stride_w, const int pad_w,
-                                       DType* out_data, bool getAvg = false) {
+                                       DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Please add `const` for `bool` and `int`.

File: src/operator/nn/pool.cuh [hunk 2]
```diff
@@ -212,23 +213,22 @@ template <typename DType>
...
+    int wstart = pw * stride_w - pad_w;
+    int wend = min(wstart + kernel_w, width + pad_w);
+    const int pool_size = (getAvg? (wend - wstart) : 1);
+    wstart = max(wstart, 0);
+    wend = min(wend, width);
+    DType sum = 0;
+    const DType* out_slice = in_data + (n * channels + c) * width;
     for (int w = wstart; w < wend; ++w) {
-      sum += out_slice[w];
+      sum += mshadow_op::power::Map(out_slice[w], DType(p_value));
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Can `DType` be float? This may have a performance impact on trivial case when `p=1`. Please provide benchmark results of the op using this function:
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/test_utils.py#L1133

File: src/operator/nn/pool.cuh [hunk 3]
```diff
@@ -243,30 +243,29 @@ __global__ void pool_sum_2d_gpu_kernel(const int nthreads, const DType* in_data,
                                        const int kernel_h, const int kernel_w,
                                        const int stride_h, const int stride_w,
                                        const int pad_h, const int pad_w,
-                                       DType* out_data, bool getAvg = false) {
+                                       DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Please add `const` for `bool` and `int`.

File: src/operator/nn/pool.cuh [hunk 4]
```diff
@@ -282,37 +281,36 @@ __global__ void pool_sum_3d_gpu_kernel(const int nthreads, const DType* in_data,
                                        const int kernel_h, const int kernel_w,
                                        const int stride_d, const int stride_h, const int stride_w,
                                        const int pad_d, const int pad_h, const int pad_w,
-                                       DType* out_data, bool getAvg = false) {
+                                       DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Add const.

File: src/operator/nn/pool.cuh [hunk 5]
```diff
@@ -484,32 +482,38 @@ __global__ void unpool_max_3d_gpu_kernel(const int nthreads, const DType* out_gr
  */
 template<typename DType>
 __global__ void unpool_sum_1d_gpu_kernel(const int nthreads, const DType* out_grad,
+                                         const DType* in_data, const DType* out_data,
                                          const int channels, const int width,
                                          const int pooled_width, const int kernel_w,
                                          const int stride_w, const int pad_w,
-                                         DType* in_grad, bool isAvg = false) {
+                                         DType* in_grad, bool isAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Add const.

File: src/operator/nn/pool.cuh [hunk 6]
```diff
@@ -519,41 +523,48 @@ __global__ void unpool_sum_1d_gpu_kernel(const int nthreads, const DType* out_gr
...
 template<typename DType>
 __global__ void unpool_sum_2d_gpu_kernel(const int nthreads, const DType* out_grad,
+                                         const DType* in_data, const DType* out_data,
                                          const int channels, const int height, const int width,
                                          const int pooled_height, const int pooled_width,
                                          const int kernel_h, const int kernel_w,
                                          const int stride_h, const int stride_w,
                                          const int pad_h, const int pad_w,
-                                         DType* in_grad, bool isAvg = false) {
+                                         DType* in_grad, bool isAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Add const.

File: src/operator/nn/pool.cuh [hunk 7]
```diff
@@ -563,31 +574,35 @@ __global__ void unpool_sum_2d_gpu_kernel(const int nthreads, const DType* out_gr
...
+                                         const DType* in_data, const DType* out_data,
                                          const int channels, const int depth, const int height,
                                          const int width, const int pooled_depth,
                                          const int pooled_height, const int pooled_width,
                                          const int kernel_d, const int kernel_h,
                                          const int kernel_w, const int stride_d, const int stride_h,
                                          const int stride_w, const int pad_d, const int pad_h,
-                                         const int pad_w, DType* in_grad, bool isAvg = false) {
+                                         const int pad_w, DType* in_grad, bool isAvg = false,
+                                         int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Add const.

File: src/operator/nn/pool.cuh [hunk 8]
```diff
@@ -621,12 +640,13 @@ __global__ void unpool_sum_3d_gpu_kernel(const int nthreads, const DType* out_gr
...
  * \param req_type operator request type, only support kWriteTo for now
  * \param out_data pointer of the output tensor data in the format of NCW, NCHW, or NCDHW
+ * \param p_value value of p for Lp pooling
  */
 template<typename DType>
 inline void pool(mshadow::Stream<gpu>* s, const DType* in_data, const TShape& ishape,
                  const TShape& oshape, const TShape& kernel, const TShape& pad,
                  const TShape& stride, const int pool_type, OpReqType req_type,
-                 DType* out_data) {
+                 DType* out_data, int p_value = 2) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: 1. Add const.
2. Why default to 2?

File: src/operator/nn/pool.cuh [hunk 9]
```diff
@@ -730,12 +774,13 @@ inline void pool(mshadow::Stream<gpu>* s, const DType* in_data, const TShape& is
...
  * \param req_type operator request type: kNullOp, kNullWriteInplace, kNullWriteTo, kNullAddTo
  * \param in_grad pointer of the gradient of the operator's input tensor
+ * \param p_value value of p for Lp pooling
  */
 template<typename DType>
 inline void unpool(mshadow::Stream<gpu>* s, const DType* out_grad, const DType* in_data,
                    const DType* out_data, const TShape& ishape, const TShape& oshape,
                    const TShape& kernel, const TShape& pad, const TShape& stride,
-                   const int pool_type, OpReqType req_type, DType* in_grad) {
+                   const int pool_type, OpReqType req_type, DType* in_grad, int p_value = 2) {
```
Comment (reminisce) [inline src/operator/nn/pool.cuh]: Add const.

File: src/operator/nn/pool.h [hunk 10]
```diff
@@ -214,7 +215,7 @@ inline void pool_max_3d_cpu(const DType* in_data, const TShape& ishape, const TS
 template<typename DType>
 inline void pool_sum_1d_cpu(const DType* in_data, const TShape& ishape, const TShape& oshape,
                             const TShape& kernel, const TShape& pad, const TShape& stride,
-                            DType* out_data, bool getAvg = false) {
+                            DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.h]: Add const.

File: src/operator/nn/pool.h [hunk 11]
```diff
@@ -249,7 +250,7 @@ inline void pool_sum_1d_cpu(const DType* in_data, const TShape& ishape, const TS
 template<typename DType>
 inline void pool_sum_2d_cpu(const DType* in_data, const TShape& ishape, const TShape& oshape,
                             const TShape& kernel, const TShape& pad, const TShape& stride,
-                            DType* out_data, bool getAvg = false) {
+                            DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.h]: Add const.

File: src/operator/nn/pool.h [hunk 12]
```diff
@@ -292,7 +294,7 @@ inline void pool_sum_2d_cpu(const DType* in_data, const TShape& ishape, const TS
 template<typename DType>
 inline void pool_sum_3d_cpu(const DType* in_data, const TShape& ishape, const TShape& oshape,
                             const TShape& kernel, const TShape& pad, const TShape& stride,
-                            DType* out_data, bool getAvg = false) {
+                            DType* out_data, bool getAvg = false, int p_value = 1) {
```
Comment (reminisce) [inline src/operator/nn/pool.h]: Add const.

File: tests/python/gpu/test_operator_gpu.py [hunk 13]
```diff
@@ -745,7 +745,7 @@ def test_pooling_with_type():
 @with_seed()
 def test_pooling_versions():
     def test_pooling_versions_helper(pool_op_list, data, kernel, pool_type, pad, stride,
-                                     pooling_convention='valid', global_pool=False):
+                                     pooling_convention='valid', global_pool=False, p_value=2):
```
Comment (reminisce) [inline tests/python/gpu/test_operator_gpu.py]: why default to 2?

Comment (haojin2) [reply tests/python/gpu/test_operator_gpu.py]: The most commonly used case for Lp pooling is p=2, for all other types of pooling this value is not used at all, so there's no side effect on other types of pooling with setting p_value to be 2.

File: src/operator/nn/pool_utils.h [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+#define MXNET_OPERATOR_NN_POOL_UTILS_H_
+
+#include "../mshadow_op.h"
+
+namespace mxnet {
+namespace op {
+
+template<typename DType, int p>
+struct a_to_p {
+  static MSHADOW_XINLINE DType Map(DType a) {
```
Comment (reminisce) [inline src/operator/nn/pool_utils.h]: nit: always use `const` if possible.

File: src/operator/nn/pool_utils.h [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+struct lp_grad<DType, 2> {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad * in_data / out_data;
+  }
+};
+
+template<typename DType>
+struct lp_grad<DType, 3> {
+  static MSHADOW_XINLINE DType Map(DType grad, DType in_data, DType out_data) {
+    return grad * in_data * in_data / out_data / out_data;
```
Comment (reminisce) [inline src/operator/nn/pool_utils.h]: Floating number division is slow. You can reduce 2 divisions into 1 and it would be faster.

File: src/operator/nn/pooling.cc [hunk 16]
```diff
@@ -344,11 +347,18 @@ Three pooling options are supported by ``pool_type``:
...
 - **sum**: sum pooling
+- **lp**: Lp pooling
 
 For 3-D pooling, an additional *depth* dimension is added before
 *height*. Namely the input data will have shape *(batch_size, channel, depth,
 height, width)*.
 
+Notes on Lp pooling:
+
+Lp pooling was first introduced by this paper: https://arxiv.org/pdf/1204.3968.pdf.
```
Comment (reminisce) [inline src/operator/nn/pooling.cc]: It's better to give a math equation here. You can adopt the format here:
https://github.com/apache/incubator-mxnet/blob/d2a856a3a2abb4e72edc301b8b821f0b75f30722/src/operator/tensor/elemwise_binary_scalar_op_extended.cc#L89

File: src/operator/nn/pool_utils.h (hunk 15 above)
Comment (haojin2) [reply src/operator/nn/pool_utils.h]: Do you mean something like ```return grad * in_data * in_data / (out_data * out_data)```?

File: src/operator/nn/pooling.cc (hunk 16 above)
Comment (haojin2) [reply src/operator/nn/pooling.cc]: Done

File: src/operator/nn/pooling.cc [hunk 17]
```diff
@@ -92,6 +92,9 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
                          std::vector<TShape> *out_shape) {
   const PoolingParam &param = nnvm::get<PoolingParam>(attrs.parsed);
   CHECK_EQ(in_shape->size(), 1U);
+  if (param.pool_type == pool_enum::kLpPooling) {
+    CHECK(param.p_value.has_value());
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/nn/pooling.cc]: We'd better not do this as this may affect Jun's work. If we set p_value to be defaulting to 1 instead of optional, the generated json file for this symbol would contain an extra field.

File: tests/python/gpu/test_operator_gpu.py [hunk 18]
```diff
@@ -769,140 +769,161 @@ def test_pooling_versions_helper(pool_op_list, data, kernel, pool_type, pad, str
...
     test_2d_pooling('avg')
     test_2d_pooling('sum')
+    test_2d_pooling('lp', p_value=1)
+    test_2d_pooling('lp', p_value=2)
 
     test_3d_pooling('max')
     test_3d_pooling('avg')
     test_3d_pooling('sum')
+    test_3d_pooling('lp', p_value=1)
+    test_3d_pooling('lp', p_value=2)
```
[the remark this answers is not in the document]
Comment (haojin2) [reply tests/python/gpu/test_operator_gpu.py]: Added

**Comment by eric-haibin-lin** [comment]
Merging it since it's approved by 2 committers.
